### PR TITLE
Improve sample distribution errors

### LINKED ISSUE #48: Sample distribution error
## Context & versions

To sample our distributions, we need to define a failure parameter:
```
        let epsilon_fail: u64 = 1 << sec_param;
        let k = log_base2(n.saturating_mul(epsilon_fail));
```
Unfortunately we cannot used lambda as we may raise an integer overflow, hence we need to choose a reasonable parameter.

### PR DESCRIPTION
**Pull request by tolikzinovyev**
## Content

Improves the sample distribution errors. For the uniform distribution, the error becomes 2^-88 for reasonable n. For bernoulli, it is in the order of 2^-64. In reality, in one of the tests it is 2^-55 but this is the best we can do because similar errors are already introduced in parameter derivation.

## Pre-submit checklist

- Branch
    - [x] Tests are provided (if possible)
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
- PR
    - [x] No clippy warnings in the CI
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested
- Documentation
    - [x] Update README file (if relevant)
    - [x] Update documentation website (if relevant)

## Issue(s)

Closes #48

### CODE REVIEW

**Review by rrtoledo — APPROVED** [review]
LGTM

I would just remove the hash type and be more precise in one comment

File: src/utils/sample.rs [hunk 1]
```diff
@@ -1,37 +1,50 @@
...
-            )
-        }
-        let epsilon_fail: u64 = 1 << 40; // TODO: update
-        let k = log_base2(n.saturating_mul(epsilon_fail));
-        let new_n: u64 = 1 << k;
-        let d = new_n.div_ceil(n);
-        let i = mod_power_of_2(hash, new_n);
+//! Sample functions for Alba primitives
+
+type Hash = [u8; 16];
```
[the remark this answers is not in the document]
Comment (tolikzinovyev) [reply src/utils/sample.rs]: It is used multiple times: in `sample_uniform()` and `sample_bernoulli()`. I renamed to `Seed`, what do you think?

File: src/utils/sample.rs [hunk 2]
```diff
@@ -1,37 +1,50 @@
...
+/// Checks that N >= M at compile time.
+fn truncate_array<const N: usize, const M: usize>(arr: &[u8; N]) -> &[u8; M] {
+    const {
+        assert!(N >= M, "provided array is too small");
+    }
+    arr[..M].try_into().unwrap()
+}
+
+/// Shorten a `hash` array by keeping the first 16 bytes and dropping the rest.
+fn to_hash<const N: usize>(hash: &[u8; N]) -> &Hash {
```
[the remark this answers is not in the document]
Comment (tolikzinovyev) [reply src/utils/sample.rs]: `truncate_array` is used multiple times. `to_hash` is indeed not necessary, I removed it.

File: src/utils/sample.rs [hunk 3]
```diff
@@ -1,37 +1,50 @@
...
+pub(crate) fn sample_uniform<const N: usize>(hash: &[u8; N], n: u64) -> Option<u64> {
+    let hash = to_hash(hash);
+
+    // Computes the integer representation of hash modulo n when n is not a
+    // power of two.
+    fn mod_not_power_of_2(hash: &Hash, n: u64) -> Option<u64> {
+        let n = u128::from(n);
 
-        if i >= d.saturating_mul(n) {
+        // Equals 2^128 / n since n is not a power of two.
```
[the remark this answers is not in the document]
Comment (tolikzinovyev) [reply src/utils/sample.rs:27]: There are no precision losses here. We want to calculate floor(2^128 / n). Should I add "floor" in the comment?

File: src/utils/sample.rs [hunk 4]
```diff
@@ -41,34 +54,88 @@ pub(crate) fn sample_uniform(hash: &[u8], n: u64) -> Option<u64> {
...
+
+    // We find an approximation x/y of q such that 0 <= q - x/y <= ε, where y = 2^128
+    // and ε is a negligible number. Here ε is in the order of 2^-64 since the f64
+    // calculation below has 64-bit precision. This is the best we can do since q
+    // (which is f64) can already be 2^-64 away from the intended value.
+
+    // Equals 2^128. We don't use f64::powi() since that function is non-deterministic.
+    // See its documentation.
+    const Y: f64 = (u128::pow(2, 127) as f64) * 2.0;
+    // It is possible that Y*q is bigger than the maximum value of u128. In that case,
```
[the remark this answers is not in the document]
Comment (tolikzinovyev) [reply src/utils/sample.rs:67]: Maybe not, but the contract of this function doesn't restrict `q` to be less than 1.

File: src/utils/sample.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (rrtoledo) [reply src/utils/sample.rs]: Thanks!

File: src/utils/sample.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (rrtoledo) [reply src/utils/sample.rs]: Sounds good. I just thought we could use [u8; 16] directly but it is used in a couple of places sa you said

File: src/utils/sample.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (rrtoledo) [reply src/utils/sample.rs:27]: Ah, yeah adding the floor in the comment would help :)

I was starting to write that technically, if not for the type of course, it is not equal as we have `u128::MAX / n = (2^128-1) / n`.

[the remark this answers is not in the document]
Comment (rrtoledo) [reply src/utils/sample.rs:27]: Thanks!

[the remark this answers is not in the document]
Comment (tolikzinovyev) [reply src/utils/sample.rs]: Added "floor". Yeah, in this case the equality holds.

**Review by curiecrypt — APPROVED** [review]
LGTM

**Review by rrtoledo — APPROVED** [review]
LGTM

**Pull request merged**